src/arena.rs

```diff
@@ -0,0 +1,361 @@
+    leaf_nodes: InnerArena<LeafNode<K, V>>,
+    internal_nodes: InnerArena<InternalNode<K, V>>,
+    arena_id: usize,
+}
+
+unsafe impl<K: Send, V: Send> Send for Arena<K, V> {}
+unsafe impl<K, V> Sync for Arena<K, V> {}
+
+impl<K, V> std::fmt::Debug for Arena<K, V> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Arena").finish()
+    }
+}
+
+impl<K, V> Default for Arena<K, V> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<K, V> Arena<K, V> {
+    /// Construct a new arena.
+    pub fn new() -> Self {
+        let arena_id = {
+            static ID_COUNTER: AtomicUsize = AtomicUsize::new(0);
+            ID_COUNTER.fetch_add(1, Ordering::SeqCst)
+        };
+        Arena {
+            leaf_nodes: InnerArena::new(),
+            internal_nodes: InnerArena::new(),
+            arena_id,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn id(&self) -> usize {
+        self.arena_id
+    }
+
+    pub(crate) fn allocate_leaf_node(&mut self) -> Id<LeafNode<K, V>> {
+        self.leaf_nodes.allocate()
+    }
+
+    /// Get a pointer to the leaf node associated with the given `id`.
+    ///
+    /// # Safety
+    ///
+    /// The leaf node associated with the given `id` must be allocated.
+    ///
+    /// Does not check that `id` is from this arena, so if it is from a
+    /// different arena, this could lead to use-after-free and
+    /// out-of-bounds-access bugs!
+    ///
+    /// Does not check borrows!
+    pub(crate) unsafe fn leaf_node(&self, id: Id<LeafNode<K, V>>) -> *mut LeafNode<K, V> {
+        self.leaf_nodes.get(id)
+    }
+
+    pub(crate) unsafe fn deallocate_leaf_node(&mut self, id: Id<LeafNode<K, V>>) {
+        self.leaf_nodes.deallocate(id);
+    }
+
+    pub(crate) fn allocate_internal_node(&mut self) -> Id<InternalNode<K, V>> {
+        self.internal_nodes.allocate()
+    }
+
+    /// Get a pointer to the leaf node associated with the given `id`.
+    ///
+    /// # Safety
+    ///
+    /// The leaf node associated with the given `id` must be allocated.
+    ///
+    /// Does not check that `id` is from this arena, so if it is from a
+    /// different arena, this could lead to use-after-free and
+    /// out-of-bounds-access bugs!
+    ///
+    /// Does not check borrows!
+    pub(crate) unsafe fn internal_node(
+        &self,
+        id: Id<InternalNode<K, V>>,
+    ) -> *mut InternalNode<K, V> {
+        self.internal_nodes.get(id)
+    }
+
+    pub(crate) unsafe fn deallocate_internal_node(&mut self, id: Id<InternalNode<K, V>>) {
+        self.internal_nodes.deallocate(id);
+    }
+
+    pub(crate) fn has_empty_capacity(&self) -> bool {
+        self.leaf_nodes.capacity() == 0 && self.internal_nodes.capacity() == 0
+    }
+}
+
+#[derive(Copy, Clone)]
+struct Free {
+    // `u32::MAX` is a sentinal for "none".
+    next_free: u32,
+}
+
+union MaybeFree<T> {
+    free: Free,
+    allocated: ManuallyDrop<MaybeUninit<UnsafeCell<T>>>,
```

Comment:
This can just be `ManuallyDrop<UnsafeCell<T>>`. The `MaybeUninit` is implied by the union.